src/core/abstract.ml

```diff
@@ -9,46 +9,59 @@ let build_abstract a = match a.a_impl with
 	| Some c -> ignore(c.cl_build())
 	| None -> ()
 
-let has_direct_to uctx ab pl b =
-	List.exists (unify_to {uctx with allow_transitive_cast = false} ab pl b) ab.a_to
-
-let has_direct_from uctx ab pl a b =
-	List.exists (unify_from {uctx with allow_transitive_cast = false} ab pl a b) ab.a_from
+let find_cast_field uctx find =
+	let found = try
+		find {uctx with allow_transitive_cast = false}
+	with Not_found ->
+		find uctx
+	in
+	match found with
+		| Some value -> value
+		| None -> raise Not_found
 
-let find_field_to uctx ab pl b =
-	build_abstract ab;
-	List.find (unify_to_field uctx ab pl b) ab.a_to_field
+let find_field_to uctx a b ab pl =
+	List.find (unifies_to_field uctx a b ab pl) ab.a_to_field
 
-let find_field_from uctx ab pl a b =
-	build_abstract ab;
-	List.find (unify_from_field uctx ab pl a b) ab.a_from_field
+let find_field_from uctx a b ab pl =
+	List.find (unifies_from_field uctx a b ab pl) ab.a_from_field
 
-let find_to_from uctx f ab_left tl_left ab_right tl_right tleft tright =
-	build_abstract ab_left;
-	build_abstract ab_right;
-	if has_direct_to uctx ab_right tl_right tleft || has_direct_from uctx ab_left tl_left tright tleft then
-		raise Not_found
-	else
-		try f ab_right tl_right (fun () -> find_field_to uctx ab_right tl_right tleft)
-		with Not_found -> f ab_left tl_left (fun () -> find_field_from uctx ab_left tl_left tright tleft)
+let find_to_from uctx a b a1 tl1 a2 tl2 =
+	build_abstract a1;
+	build_abstract a2;
+	find_cast_field uctx (fun uctx ->
+		if unifies_abstracts uctx a b a1 tl1 a2 tl2 then
+			None
+		else try
+			Some((a1,tl1,(find_field_to uctx a b a1 tl1)))
+		with Not_found ->
+			Some((a2,tl2,(find_field_from uctx a b a2 tl2)))
+	)
 
-let find_to uctx ab pl b =
+let find_to uctx b ab pl =
```

Comment:
Why?

Reply:
Consistency of arguments between all abstract unify methods, as said in description of PR.
Currently in some places it one way, in others - another.